### ISSUE TITLE
Binary crossentropy gives NANs

### ISSUE THREAD

**Issue by Botev** [Lasagne/Lasagne#436]
So I've tried to train a convnet from scratch and encountered the problem of the binary_crossentropy giving nans all over the place at the start. Unfortunately for an autoencoder could not to force it to go away. Clipping the predictions to [1e-10,1-1e-10] works, but was wondering if this should not be built in?

**Comment by f0k** [comment]
> Clipping the predictions to [1e-10,1-1e-10] works, but was wondering if this should not be built in?

This is definitely something we should leave up to the researcher, we cannot clip anything by default. We could of course add a `clip` keyword to `binary_crossentropy`, but `binary_crossentropy(preds, targets, clip=1e-10)` doesn't read any easier than `binary_crossentropy(T.clip(preds, -1e-10, 1e-10), targets)`, so it would just complicate our code. Besides, there are many possible reasons and many possible solutions for NaNs: You might also be successful by lowering your learning rate, changing the weight initialization, normalizing/scaling your inputs, adding/changing regularization terms... or maybe Theano misses some numerical stabilization optimization, [as for the categorical crossentropy with one-hot targets](https://github.com/Theano/Theano/issues/3162).

**Comment by jiangnanhugo** [comment]
this problem is currently observed by our model with `binary_crossentropy` loss, and we have to clip its input in the code.

**Comment by Sentient07** [comment]
@jiangnanhugo did you try lowering your learning rate as Jan suggested ? What is your cost? (Mostly, `NaNs` happen when the cost value is too high to be representable)

**Comment by jiangnanhugo** [comment]
@Sentient07 thanks for your advice. Using smaller learning rate is helpful, while I wouder is this just ease the `NaN` problem but not solve it?

**Comment by f0k** [comment]
I would assume this is not due to Lasagne or Theano. Lowering the learning rate can be a valid solution, but clipping the predictions can work as well. Also see http://deeplearning.net/software/theano/tutorial/nan_tutorial.html for some hints on how to find where NaNs occur.
